### ISSUE TITLE
addNotificationResponseReceivedListener() is not called when the app opens from killed state

### ISSUE THREAD

**Issue by nsdub** [expo/expo#12954]
### Summary

Issue: In the Notifications API, `addNotificationResponseReceivedListener()` is not called when the app opens from a killed state. This matters because we can't use the data inside the push notification to intelligently navigate users to the appropriate screen.

Expected behavior: `addNotificationResponseReceivedListener()` fires when tapping on a PN when the app is backgrounded or closed.

Actual behavior: `addNotificationResponseReceivedListener()` fires when tapping on a PN **_ONLY_** when the app is backgrounded.

Notes: I've not confirmed this bug on iOS (I don't have access to an iPhone).

Related issues: Similar problems have been reported before. Here are a few examples: [https://github.com/expo/expo/issues/12312](https://github.com/expo/expo/issues/12312), [https://github.com/expo/expo/issues/12228](https://github.com/expo/expo/issues/12228), [https://github.com/expo/expo/issues/11933](https://github.com/expo/expo/issues/11933), [https://github.com/expo/expo/issues/11343](https://github.com/expo/expo/issues/11343)

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android

### SDK Version (managed workflow only)

41

### Environment

``` 
 Expo CLI 4.4.7 environment info:
    System:
      OS: macOS 11.2.3
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 15.12.0 - ~/.nvm/versions/node/v15.12.0/bin/node
      npm: 7.6.3 - ~/.nvm/versions/node/v15.12.0/bin/npm
    SDKs:
      iOS SDK:
        Platforms: iOS 14.2, DriverKit 20.0, macOS 11.0, tvOS 14.2, watchOS 7.1
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7199119
      Xcode: 12.2/12B45b - /usr/bin/xcodebuild
    npmPackages:
      expo: ^41.0.1 => 41.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-41.0.0.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.4.7
    Expo Workflow: managed
```

### Reproducible demo or steps to reproduce from a blank project

https://github.com/nsdub/addNotificationResponseReceivedListenerBug

**Comment by cruzach** [comment]
Can you provide a full repro via Snack?

**Comment by nsdub** [comment]
Hey Charlie -- 

There's a [snack bug](https://github.com/expo/expo/issues/12833 ) that impacts the ability to send PNs, so I made a [public repo here](https://github.com/nsdub/addNotificationResponseReceivedListenerBug). 

The app tells you your push token. Then just use the Expo PN Tool to fire off a PN while the app is closed and you oughta be able to reproduce the bug. Again, `addNotificationResponseReceivedListener` works while the app is foregrounded or backgrounded, but not closed.

@cruzach

**Comment by nsdub** [comment]
@cruzach is there other info from me that would be useful in diagnosing this?

**Comment by nsdub** [comment]
We got this resolved. The issue appears to have been that `addNotificationReceivedListener` and `addNotificationResponseReceivedListener` are _very_ sensitive to where they are placed in the app loading stack. 

We originally had them fire in the `componentDidMount` in the first component loaded from `App.js`. That wasn't good enough. It wasn't until we made the listeners the first functions fired in `componentDidMount` **in** `App.js` that they behaved as expected.
